File: src/storage/schema.py
TAB_TRANSACTIONS = "transactions"
TAB_DAILY_BRIEF = "daily_brief"
TAB_SUBSCRIBERS = "subscribers"
TAB_ANALYSIS_LOG = "analysis_log"
TAB_SYSTEM_LOG = "system_log"

ALL_TABS = [
    TAB_TRANSACTIONS,
    TAB_DAILY_BRIEF,
    TAB_SUBSCRIBERS,
    TAB_ANALYSIS_LOG,
    TAB_SYSTEM_LOG,
]

TRANSACTIONS_HEADERS = [
    "raw_response_hash",
    "hash",
    "timestamp",
    "blockchain",
    "symbol",
    "amount",
    "amount_usd",
    "from_address",
    "from_owner_type",
    "from_owner",
    "to_address",
    "to_owner_type",
    "to_owner",
    "created_at",
]

DAILY_BRIEF_HEADERS = [
    "date",
    "summary",
    "top_transactions",
    "total_volume_usd",
    "alert_count",
    "created_at",
    "highlights",
    "signal_themes",
    "note",
    "input_fingerprint",
]

SUBSCRIBERS_HEADERS = [
    "chat_id",
    "username",
    "status",
    "watchlist_coins",
    "created_at",
    "updated_at",
    "last_brief_at",
]

ANALYSIS_LOG_HEADERS = [
    "prompt_hash",
    "task",
    "prompt_version",
    "prompt",
    "response",
    "model",
    "model_id",
    "tokens_used",
    "tokens_in",
    "tokens_out",
    "cost_usd",
    "latency_ms",
    "status",
    "created_at",
]

SYSTEM_LOG_HEADERS = [
    "run_id",
    "run_type",
    "status",
    "started_at",
    "finished_at",
    "transactions_count",
    "errors",
    "details",
]

TAB_HEADERS = {
    TAB_TRANSACTIONS: TRANSACTIONS_HEADERS,
    TAB_DAILY_BRIEF: DAILY_BRIEF_HEADERS,
    TAB_SUBSCRIBERS: SUBSCRIBERS_HEADERS,
    TAB_ANALYSIS_LOG: ANALYSIS_LOG_HEADERS,
    TAB_SYSTEM_LOG: SYSTEM_LOG_HEADERS,
}

# --- New tabs (TRACK 2) ---

TAB_WATCHED_ADDRESSES = "watched_addresses"
TAB_ADDRESS_ACTIVITY = "address_activity"
TAB_TG_WHALE_EVENTS = "tg_whale_events"
TAB_SIGNALS = "signals"
TAB_WEEKLY_TREND = "weekly_trend"
TAB_USER_INTERESTS = "user_interests"
TAB_CURATED_WALLETS = "curated_wallets"
TAB_WALLET_ALIASES = "wallet_aliases"
TAB_WATCHLIST_OVERRIDES = "watchlist_overrides"
TAB_NEWS_FEED = "news_feed"
TAB_MARKET_SNAPSHOTS = "market_snapshots"
TAB_WALLET_ACTIVITY_SNAPSHOTS = "wallet_activity_snapshots"
TAB_WHALE_STORIES = "whale_stories"
TAB_BROADCAST_LOG = "broadcast_log"
TAB_LLM_BUDGET_LOG = "llm_budget_log"
TAB_CURATED_WALLET_BALANCES = "curated_wallet_balances"
TAB_CHANNEL_HEALTH = "channel_health"
TAB_SERVICE_HEALTH = "service_health"

WATCHED_ADDRESSES_HEADERS = [
    "address", "chain", "category", "label", "source",
    "confidence", "enabled", "added_at", "notes",
]

ADDRESS_ACTIVITY_HEADERS = [
    "tx_hash", "chain", "block_time", "watched_address", "direction",
    "counterparty", "counterparty_category", "token", "amount_token",
    "amount_usd", "collected_at",
]

TG_WHALE_EVENTS_HEADERS = [
    "tg_msg_id", "tg_date", "blockchain", "symbol", "amount",
    "amount_usd", "from_owner_type", "from_owner", "to_owner_type",
    "to_owner", "raw_text", "parsed_confidence", "collected_at",
]

SIGNALS_HEADERS = [
    "signal_id", "created_at", "rule", "severity", "score",
    "confidence", "source", "evidence_tx_hashes", "window_start",
    "window_end", "summary", "extra_json",
]

WEEKLY_TREND_HEADERS = [
    "week_start", "category", "chain", "net_flow_usd",
    "event_count", "unique_addresses", "created_at",
]

USER_INTERESTS_HEADERS = [
    "chat_id", "dimension", "value", "weight", "source",
    "updated_at",
]

CURATED_WALLETS_HEADERS = [
    "id", "chain", "address", "owner_label", "owner_category",
    "owner_subcategory", "approx_balance", "tier", "source_ref",
    "source_url", "note", "entity_id", "is_representative",
    "narrative_tags", "display_priority", "is_active",
    "created_at", "updated_at",
]

WALLET_ALIASES_HEADERS = [
    "canonical_id", "alias_id", "chain", "address", "label", "note",
]

WATCHLIST_OVERRIDES_HEADERS = [
    "wallet_id", "enabled", "actor", "reason", "updated_at",
]

NEWS_FEED_HEADERS = [
    "id", "source", "title", "summary", "url",
    "published_at", "language", "tags", "fetched_at", "hash",
    # last_seen_at = when we most recently observed this article in ANY RSS poll,
    # including dedup hits. Distinct from fetched_at (first insertion time) so the
    # dashboard can tell "pipeline is polling" apart from "new article arrived".
    "last_seen_at",
]

MARKET_SNAPSHOTS_HEADERS = [
    "ts", "symbol", "binance_usd", "upbit_krw", "bitflyer_jpy",
    "kraken_eur", "krw_premium_pct", "jpy_premium_pct", "eur_premium_pct",
]

WALLET_ACTIVITY_SNAPSHOTS_HEADERS = [
    "ts", "wallet_id", "chain", "balance", "balance_delta_24h",
    "inflow_24h", "outflow_24h", "tx_count_24h", "source",
]

WHALE_STORIES_HEADERS = [
    "id", "signal_id", "wallet_id", "title", "body_ko",
    "body_en", "impact_score", "published_at", "source_signal_ts",
]

BROADCAST_LOG_HEADERS = [
    "ts", "kind", "dedup_key", "chat_id", "message_id", "status", "error",
]

LLM_BUDGET_LOG_HEADERS = [
    "ts",
    "month_key",
    "pipeline",
    "model_id",
    "tokens_in",
    "tokens_out",
    "cost_usd",
    "cumulative_cost_usd",
    "decision",
]

CURATED_WALLET_BALANCES_HEADERS = [
    "wallet_id",
    "chain",
    "address",
    "owner_label",
    "owner_category",
    "approx_balance",
    "source_ref",
    "source_url",
    "note",
    "is_active",
    "updated_at",
]

CHANNEL_HEALTH_HEADERS = [
    "ts",
    "chat_id",
    "title",
    "username",
    "member_count",
    "status",
    "error",
]

SERVICE_HEALTH_HEADERS = [
    "ts",
    "service",
    "component",
    "status",
    "heartbeat_key",
    "details",
    "error",
]

ALL_TABS.extend([
    TAB_WATCHED_ADDRESSES,
    TAB_ADDRESS_ACTIVITY,
    TAB_TG_WHALE_EVENTS,
    TAB_SIGNALS,
    TAB_WEEKLY_TREND,
    TAB_USER_INTERESTS,
    TAB_CURATED_WALLETS,
    TAB_WALLET_ALIASES,
    TAB_WATCHLIST_OVERRIDES,
    TAB_NEWS_FEED,
    TAB_MARKET_SNAPSHOTS,
    TAB_WALLET_ACTIVITY_SNAPSHOTS,
    TAB_WHALE_STORIES,
    TAB_BROADCAST_LOG,
    TAB_LLM_BUDGET_LOG,
    TAB_CURATED_WALLET_BALANCES,
    TAB_CHANNEL_HEALTH,
    TAB_SERVICE_HEALTH,
])

TAB_HEADERS.update({
    TAB_WATCHED_ADDRESSES: WATCHED_ADDRESSES_HEADERS,
    TAB_ADDRESS_ACTIVITY: ADDRESS_ACTIVITY_HEADERS,
    TAB_TG_WHALE_EVENTS: TG_WHALE_EVENTS_HEADERS,
    TAB_SIGNALS: SIGNALS_HEADERS,
    TAB_WEEKLY_TREND: WEEKLY_TREND_HEADERS,
    TAB_USER_INTERESTS: USER_INTERESTS_HEADERS,
    TAB_CURATED_WALLETS: CURATED_WALLETS_HEADERS,
    TAB_WALLET_ALIASES: WALLET_ALIASES_HEADERS,
    TAB_WATCHLIST_OVERRIDES: WATCHLIST_OVERRIDES_HEADERS,
    TAB_NEWS_FEED: NEWS_FEED_HEADERS,
    TAB_MARKET_SNAPSHOTS: MARKET_SNAPSHOTS_HEADERS,
    TAB_WALLET_ACTIVITY_SNAPSHOTS: WALLET_ACTIVITY_SNAPSHOTS_HEADERS,
    TAB_WHALE_STORIES: WHALE_STORIES_HEADERS,
    TAB_BROADCAST_LOG: BROADCAST_LOG_HEADERS,
    TAB_LLM_BUDGET_LOG: LLM_BUDGET_LOG_HEADERS,
    TAB_CURATED_WALLET_BALANCES: CURATED_WALLET_BALANCES_HEADERS,
    TAB_CHANNEL_HEALTH: CHANNEL_HEALTH_HEADERS,
    TAB_SERVICE_HEALTH: SERVICE_HEALTH_HEADERS,
})

File: apps/dashboard/lib/schema.ts
export const TAB_TRANSACTIONS = "transactions" as const;
export const TAB_DAILY_BRIEF = "daily_brief" as const;
export const TAB_SIGNALS = "signals" as const;
export const TAB_SYSTEM_LOG = "system_log" as const;
export const TAB_SUBSCRIBERS = "subscribers" as const;
export const TAB_TG_WHALE_EVENTS = "tg_whale_events" as const;
export const TAB_CURATED_WALLETS = "curated_wallets" as const;
export const TAB_WATCHED_ADDRESSES = "watched_addresses" as const;
export const TAB_WALLET_ALIASES = "wallet_aliases" as const;
export const TAB_WATCHLIST_OVERRIDES = "watchlist_overrides" as const;
export const TAB_NEWS_FEED = "news_feed" as const;

export const TRANSACTIONS_HEADERS = [
  "raw_response_hash",
  "hash",
  "timestamp",
  "blockchain",
  "symbol",
  "amount",
  "amount_usd",
  "from_address",
  "from_owner_type",
  "from_owner",
  "to_address",
  "to_owner_type",
  "to_owner",
  "created_at",
] as const;

export const DAILY_BRIEF_HEADERS = [
  "date",
  "summary",
  "top_transactions",
  "total_volume_usd",
  "alert_count",
  "created_at",
  "highlights",
  "signal_themes",
  "note",
] as const;

export const SIGNALS_HEADERS = [
  "signal_id",
  "created_at",
  "rule",
  "severity",
  "score",
  "confidence",
  "source",
  "evidence_tx_hashes",
  "window_start",
  "window_end",
  "summary",
  "extra_json",
] as const;

export const SYSTEM_LOG_HEADERS = [
  "run_id",
  "run_type",
  "status",
  "started_at",
  "finished_at",
  "transactions_count",
  "errors",
  "details",
] as const;

export const SUBSCRIBERS_HEADERS = [
  "chat_id",
  "username",
  "status",
  "watchlist_coins",
  "created_at",
  "updated_at",
  "last_brief_at",
] as const;

export const TG_WHALE_EVENTS_HEADERS = [
  "tg_msg_id",
  "tg_date",
  "blockchain",
  "symbol",
  "amount",
  "amount_usd",
  "from_owner_type",
  "from_owner",
  "to_owner_type",
  "to_owner",
  "raw_text",
  "parsed_confidence",
  "collected_at",
] as const;

export const CURATED_WALLETS_HEADERS = [
  "id",
  "chain",
  "address",
  "owner_label",
  "owner_category",
  "owner_subcategory",
  "approx_balance",
  "tier",
  "source_ref",
  "source_url",
  "note",
  "is_active",
  "created_at",
  "updated_at",
] as const;

export const WATCHED_ADDRESSES_HEADERS = [
  "address",
  "chain",
  "category",
  "label",
  "source",
  "confidence",
  "enabled",
  "added_at",
  "notes",
] as const;

export const WALLET_ALIASES_HEADERS = [
  "canonical_id",
  "alias_id",
  "chain",
  "address",
  "label",
  "note",
] as const;

export const WATCHLIST_OVERRIDES_HEADERS = [
  "wallet_id",
  "enabled",
  "actor",
  "reason",
  "updated_at",
] as const;

export const NEWS_FEED_HEADERS = [
  "id",
  "source",
  "title",
  "summary",
  "url",
  "published_at",
  "language",
  "tags",
  "fetched_at",
  "hash",
  // last_seen_at = most recent RSS poll that observed this article (including dedup hits).
  // Distinct from fetched_at (first insert) so the dashboard can tell
  // "pipeline is polling" apart from "new article arrived".
  "last_seen_at",
] as const;

export interface TransactionRow {
  raw_response_hash: string;
  hash: string;
  timestamp: string;
  blockchain: string;
  symbol: string;
  amount: string;
  amount_usd: string;
  from_address: string;
  from_owner_type: string;
  from_owner: string;
  to_address: string;
  to_owner_type: string;
  to_owner: string;
  created_at: string;
}

export interface DailyBriefRow {
  date: string;
  summary: string;
  top_transactions: string;
  total_volume_usd: string;
  alert_count: string;
  created_at: string;
  highlights: string;
  signal_themes: string;
  note: string;
}

export interface SignalRow {
  signal_id: string;
  created_at: string;
  rule: string;
  severity: string;
  score: string;
  confidence: string;
  source: string;
  evidence_tx_hashes: string;
  window_start: string;
  window_end: string;
  summary: string;
  extra_json: string;
}

export interface SystemLogRow {
  run_id: string;
  run_type: string;
  status: string;
  started_at: string;
  finished_at: string;
  transactions_count: string;
  errors: string;
  details: string;
}

export interface TgWhaleEventRow {
  tg_msg_id: string;
  tg_date: string;
  blockchain: string;
  symbol: string;
  amount: string;
  amount_usd: string;
  from_owner_type: string;
  from_owner: string;
  to_owner_type: string;
  to_owner: string;
  raw_text: string;
  parsed_confidence: string;
  collected_at: string;
}

export interface CuratedWalletRow {
  id: string;
  chain: string;
  address: string;
  owner_label: string;
  owner_category: string;
  owner_subcategory: string;
  approx_balance: string;
  tier: string;
  source_ref: string;
  source_url: string;
  note: string;
  is_active: string;
  created_at: string;
  updated_at: string;
}

export interface WatchedAddressRow {
  address: string;
  chain: string;
  category: string;
  label: string;
  source: string;
  confidence: string;
  enabled: string;
  added_at: string;
  notes: string;
}

export interface WalletAliasRow {
  canonical_id: string;
  alias_id: string;
  chain: string;
  address: string;
  label: string;
  note: string;
}

export interface WatchlistOverrideRow {
  wallet_id: string;
  enabled: string;
  actor: string;
  reason: string;
  updated_at: string;
}

export interface NewsFeedRow {
  id: string;
  source: string;
  title: string;
  summary: string;
  url: string;
  published_at: string;
  language: string;
  tags: string;
  fetched_at: string;
  hash: string;
  last_seen_at: string;
}

export interface SubscriberRow {
  chat_id: string;
  username: string;
  status: string;
  watchlist_coins: string;
  created_at: string;
  updated_at: string;
  last_brief_at: string;
}

export interface SheetRowMap {
  transactions: TransactionRow;
  daily_brief: DailyBriefRow;
  signals: SignalRow;
  system_log: SystemLogRow;
  subscribers: SubscriberRow;
  tg_whale_events: TgWhaleEventRow;
  curated_wallets: CuratedWalletRow;
  watched_addresses: WatchedAddressRow;
  wallet_aliases: WalletAliasRow;
  watchlist_overrides: WatchlistOverrideRow;
  news_feed: NewsFeedRow;
}

export type SheetTabName = keyof SheetRowMap;

export const TAB_HEADERS = {
  transactions: TRANSACTIONS_HEADERS,
  daily_brief: DAILY_BRIEF_HEADERS,
  signals: SIGNALS_HEADERS,
  system_log: SYSTEM_LOG_HEADERS,
  subscribers: SUBSCRIBERS_HEADERS,
  tg_whale_events: TG_WHALE_EVENTS_HEADERS,
  curated_wallets: CURATED_WALLETS_HEADERS,
  watched_addresses: WATCHED_ADDRESSES_HEADERS,
  wallet_aliases: WALLET_ALIASES_HEADERS,
  watchlist_overrides: WATCHLIST_OVERRIDES_HEADERS,
  news_feed: NEWS_FEED_HEADERS,
} as const satisfies Record<SheetTabName, readonly string[]>;

export const DASHBOARD_TABS = [
  TAB_TRANSACTIONS,
  TAB_DAILY_BRIEF,
  TAB_SIGNALS,
  TAB_SYSTEM_LOG,
  TAB_SUBSCRIBERS,
  TAB_TG_WHALE_EVENTS,
] as const;

export type DashboardTabName = (typeof DASHBOARD_TABS)[number];
